Source/SpatialGDK/Private/Utils/ComponentReader.cpp

```diff
@@ -0,0 +1,379 @@
+
+	TArray<std::uint32_t> UpdateFields;
+	UpdateFields.SetNum(Schema_GetUniqueFieldIdCount(ComponentObject));
+	Schema_GetUniqueFieldIds(ComponentObject, UpdateFields.GetData());
+
+	if (UpdateFields.Num() == 0)
+	{
+		return;
+	}
+
+	FObjectReplicator& Replicator = Channel->PreReceiveSpatialUpdate(Object);
+
+	FRepState* RepState = Replicator.RepState;
+	TArray<FRepLayoutCmd>& Cmds = Replicator.RepLayout->Cmds;
+	TArray<FHandleToCmdIndex>& BaseHandleToCmdIndex = Replicator.RepLayout->BaseHandleToCmdIndex;
+	TArray<FRepParentCmd>& Parents = Replicator.RepLayout->Parents;
+
+	bool bIsServer = NetDriver->IsServer();
+	FSpatialConditionMapFilter ConditionMap(Channel, bAutonomousProxy);
+
+	TArray<UProperty*> RepNotifies;
+
+	for (std::uint32_t FieldId : UpdateFields)
+	{
+		// FieldId is the same as rep handle
+		check((int)FieldId - 1 < BaseHandleToCmdIndex.Num());
+		const FRepLayoutCmd& Cmd = Cmds[BaseHandleToCmdIndex[FieldId - 1].CmdIndex];
+		const FRepParentCmd& Parent = Parents[Cmd.ParentIndex];
+
+		if (bIsServer || ConditionMap.IsRelevant(Parent.Condition))
+		{
+			// This swaps Role/RemoteRole as we write it
+			const FRepLayoutCmd& SwappedCmd = (!bIsServer && Parent.RoleSwapIndex != -1) ? Cmds[Parents[Parent.RoleSwapIndex].CmdStart] : Cmd;
+
+			uint8* Data = (uint8*)Object + SwappedCmd.Offset;
+
+			if (bIsInitialData || GetPropertyCount(ComponentObject, FieldId, Cmd.Property) > 0 || ClearedIds->Find(FieldId) != INDEX_NONE)
+			{
+				if (Cmd.Type == REPCMD_DynamicArray)
+				{
+					ApplyArray(ComponentObject, FieldId, Cast<UArrayProperty>(Cmd.Property), Data, SwappedCmd.Offset, Cmd.ParentIndex);
+				}
+				else
+				{
+					ApplyProperty(ComponentObject, FieldId, 0, Cmd.Property, Data, SwappedCmd.Offset, Cmd.ParentIndex);
+				}
+
+				if (Cmd.Property->GetFName() == NAME_RemoteRole)
+				{
+					// Downgrade role from AutonomousProxy to SimulatedProxy if we aren't authoritative over
+					// the client RPCs component.
+					UByteProperty* ByteProperty = Cast<UByteProperty>(Cmd.Property);
+					if (!bIsServer && !bAutonomousProxy && ByteProperty->GetPropertyValue(Data) == ROLE_AutonomousProxy)
+					{
+						ByteProperty->SetPropertyValue(Data, ROLE_SimulatedProxy);
+					}
+				}
+
+				// Parent.Property is the "root" replicated property, e.g. if a struct property was flattened
+				if (Parent.Property->HasAnyPropertyFlags(CPF_RepNotify))
+				{
+					if (Parent.RepNotifyCondition == REPNOTIFY_Always || !Cmd.Property->Identical(RepState->StaticBuffer.GetData() + SwappedCmd.Offset, Data))
+					{
+						RepNotifies.AddUnique(Parent.Property);
+					}
+				}
+			}
+		}
+	}
+
+	Channel->PostReceiveSpatialUpdate(Object, RepNotifies);
+}
+
+void ComponentReader::ApplyProperty(Schema_Object* Object, Schema_FieldId Id, std::uint32_t Index, UProperty* Property, uint8* Data, int32 Offset, uint16 ParentIndex)
+{
+	if (UStructProperty* StructProperty = Cast<UStructProperty>(Property))
```

Comment:
There is _a lot_ of branching in this code, which is on a hot code path. Is there a way of mapping a property type to some ID which we can `switch` on (which the compiler will convert into a jump table).